### ISSUE TITLE
`event` argument type

### ISSUE THREAD

**Issue by arcnmx** [ilianaw/rust-crowbar#12]
Along the same lines as #11 it would be nice to move away from `event: Value` as the input to a lambda function. A few options:

- Allowing anything `: Deserialize` would be great but would rely on inference, which may require an annotation of the type in the handler definition. Not a problem at all for `fn handler(event: Whatever, context: Context)`, but maybe slightly annoying for `|event: Whatever, context|`? Not sure if that's a real concern or not, this seems like a nice ergonomic approach.
- `Value` type could just be a newtype wrapper around `PyObject` that impls `Deserializer` so that handlers could just do: `let event = Whatever::deserialize(event)?`. Value-esque accessors to the PyObject's attributes could be supplied for anyone who wants to use it without serde and just access things by string keys.

**Comment by softprops** [comment]
+1

**Comment by softprops** [comment]
I wonder if this could be accomplished as is without a breaking change since deserialize is already implemented for value https://docs.serde.rs/serde_json/enum.Value.html

**Comment by softprops** [comment]
Using something like https://docs.serde.rs/serde_json/fn.from_value.html

**Comment by arcnmx** [comment]
True, `Value` does impl `Deserializer` already, and that is how handlers using serde need to work currently. The main advantage of the newtype approach would be that the implementation details are hidden, so that the PyObject could be passed directly without explosing the `Value` interim type (which is unnecessary extra allocations/copying/processing/etc) unless the handler wants it.

Since `Value` also impls `Deserialize`, existing handlers taking the `Value` type would work in the first option as well, however would still be a breaking change due to inference changes requiring type annotations for closures.

**Comment by softprops** [comment]
Feel like this is a worth while breaking change. The only cases where in haven't called serde_json::from_value(...) are cases where I don't need to reference the event, typically scheduled cloudwatch event crons.

The advantage of the current interface is that it gives the application control over error handling when deserializing fails. 

For me, I would be willing to fold on that for the convenience of having crowbar do the deserialization for me. This also allows for future optimization where crowbar could skip the intermediary serde_json::Value type and deserialize directly from pyobject to my deserialize type with serde.

**Comment by arcnmx** [comment]
Agreed, and if you did need to do manual error handling, you could have the handler take `serde_json::Value` anyway and it would be equivalent to the current interface.

**Comment by softprops** [comment]
+1
